From dd86629bfd1829288e6786126e396e687720e52d Mon Sep 17 00:00:00 2001
From: Daniel Kolesa <daniel@octaforge.org>
Date: Tue, 21 Dec 2021 02:46:30 +0100
Subject: [PATCH 12/15] allow specifying native sysroot to use for linkage

This allows us to get around the linker attempting to use
incompatible libs.
---
 src/bootstrap/src/core/builder/cargo.rs      | 4 ++++
 src/bootstrap/src/core/config/toml/target.rs | 3 +++
 src/bootstrap/src/lib.rs                     | 4 ++++
 3 files changed, 11 insertions(+)

diff --git a/src/bootstrap/src/core/builder/cargo.rs b/src/bootstrap/src/core/builder/cargo.rs
index 62c4d85ee..cd9431c38 100644
--- a/src/bootstrap/src/core/builder/cargo.rs
+++ b/src/bootstrap/src/core/builder/cargo.rs
@@ -265,6 +265,10 @@ impl Cargo {
             self.hostflags.arg(&arg);
         }
 
+        if let Some(sysroot) = builder.native_sysroot(target) {
+            self.rustflags.arg(&format!("-Clink-args=--sysroot={}", sysroot.display()));
+        }
+
         if let Some(target_linker) = builder.linker(target) {
             let target = crate::envify(&target.triple);
             self.command.env(format!("CARGO_TARGET_{target}_LINKER"), target_linker);
diff --git a/src/bootstrap/src/core/config/toml/target.rs b/src/bootstrap/src/core/config/toml/target.rs
index e3a8d0d3c..f893c8c4f 100644
--- a/src/bootstrap/src/core/config/toml/target.rs
+++ b/src/bootstrap/src/core/config/toml/target.rs
@@ -30,6 +30,7 @@ define_config! {
         default_linker: Option<PathBuf> = "default-linker",
         linker: Option<String> = "linker",
         split_debuginfo: Option<String> = "split-debuginfo",
+        sysroot: Option<String> = "sysroot",
         llvm_config: Option<String> = "llvm-config",
         llvm_has_rust_patches: Option<bool> = "llvm-has-rust-patches",
         llvm_filecheck: Option<String> = "llvm-filecheck",
@@ -64,6 +65,7 @@ pub struct Target {
     pub default_linker: Option<PathBuf>,
     pub linker: Option<PathBuf>,
     pub split_debuginfo: Option<SplitDebuginfo>,
+    pub sysroot: Option<PathBuf>,
     pub sanitizers: Option<bool>,
     pub profiler: Option<StringOrBool>,
     pub rpath: Option<bool>,
@@ -127,6 +129,7 @@ impl Config {
                 target.ar = cfg.ar.map(PathBuf::from);
                 target.ranlib = cfg.ranlib.map(PathBuf::from);
                 target.linker = cfg.linker.map(PathBuf::from);
+                target.sysroot = cfg.sysroot.map(PathBuf::from);
                 target.crt_static = cfg.crt_static;
                 target.wasi_root = cfg.wasi_root.map(PathBuf::from);
                 target.qemu_rootfs = cfg.qemu_rootfs.map(PathBuf::from);
diff --git a/src/bootstrap/src/lib.rs b/src/bootstrap/src/lib.rs
index 557ec1675..1163a961c 100644
--- a/src/bootstrap/src/lib.rs
+++ b/src/bootstrap/src/lib.rs
@@ -1288,6 +1288,10 @@ impl Build {
         }
     }
 
+    fn native_sysroot(&self, target: TargetSelection) -> Option<&Path> {
+        self.config.target_config.get(&target).and_then(|c| c.sysroot.as_ref()).map(|p| &**p)
+    }
+
     /// Returns the `lib` directory for the WASI target specified, if
     /// configured.
     ///
-- 
2.51.2

